쥬스메이커 [STEP 1] vetto, 레옹아범

fatherLeon opened the pull request:
@llingo
안녕하세요 링고 조금 이르지만 STEP1 나름 잘 완수했다는 생각이 들어 빠르게 PR을 보냅니다!!!
3주? 프로젝트인걸로 알고있는데 3주동안 잘 부탁드리겠습니다🙇

## 고민했던 점
### Naming
- 함수명을 처음에는 `increaseFruit`같이 길게 적었습니다.
```swift
func increaseFruit(fruit: Fruit, number: Int) {
    guard let storedFruit = self.fruits[fruit] else {
        return
    }
    self.fruits[fruit] = storedFruit + number
}
```
- instance안에 함수가 있는걸로 생각하여 나중에 고민하여 `increase`같이 증감만을 나타내는 함수명으로 바꾸었습니다. 이러한 함수명이 코드의 가독성을 헤치지 않는지 고민했습니다.
```swift
func increase(fruit: Fruit, by amount: Int) {
    guard let storedFruit = self.fruits[fruit] else {
        return
    }
    self.fruits[fruit] = storedFruit + amount
}
```
### 딕셔너리 옵셔널 처리
* 딕셔너리 사용이 익숙치 않아서인지 옵셔널 타입에서 무엇을 언래핑 해줘야할지 감을 잘 못잡았습니다.
* 또한 스위치문에서 ```switch 열거형```으로 case문을 작성할 경우 열거형의 모든 케이스가 switch문 안에 들어가 있을 경우 따로 default를 안해줘도 되듯이 딕셔너리도 마찬가지라 생각했는데 그게 아니어서 고민을 많이 했습니다.
```swift
private var fruits: [Fruit: Int] = [.strawberry: 10, .banana: 10, .kiwi: 10, .mango: 10, .pineapple: 10]

func increase(fruit: Fruit, by amount: Int) {
    guard let storedFruit = self.fruits[fruit] else {
        return
    }
    self.fruits[fruit] = storedFruit + amount
}
```
* 코드 작성을 하다보니 딕셔너리의 특정 key값을 넣어줄때는 따로 옵셔널 처리를 안해줘도 되지만 특정 key값의 value를 빼올때에는 옵셔널 처리가 필요하단것을 깨우쳐 위와같이 코드를 작성했습니다.

### 소유에 대한 의미
* 이번 프로젝트 요구사항 중에는 `JuiceMaker는 FruitStore를 소유하고 있습니다.`라는 사항이 있었습니다. 처음에는 어떠한 의미인지 모르겠어서 고민을 많이 했습니다. 
* 팀원과 상의한 결과 JuiceMaker안에 FruitStore라는 instance가 존재하면 된다고 생각하여 JuiceMaker안에서 호출을 하였습니다.
* 저희가 생각하고 있는 것이 맞는지 궁금합니다.

## 조언받고 싶은 점 && 해결하지 못한 점

### 에러처리에 대한 로직
* throw를 두번 거치는 방법에 대해 어떻게 생각하시나요?
```swift
func makeJuice(_ juice: JuiceMenu) {
    do {
        try checkFruitInJuice(juice: juice)
        useFruit(juice: juice)
        print(fruitStore.fruits)
    } catch FruitStoreError.noExistInventory {
        print("해당 과일 존재하지 않음")
    } catch FruitStoreError.lackedInventory {
        print("과일 재고 부족")
    } catch {
        print(error)
    }
}
    
private func checkFruitInJuice(_ juice: JuiceMenu) throws {
    for (fruit, num) in juice.recipe {
        try fruitStore.checkFruit(fruit: fruit, number: num)
    }
}
```
* 위는 throws를 통해서 에러가 두번 거치는 방법입니다.
* 장점 - ```makeJuice```가 에러처리를 함으로 다양한 에러에 따른 반응이 가능하다
* 단점 - ```FruitStore```의 에러가 ```checkFruitInJuice```로 이동하고 ```makeJuice```로 또 이동하기 때문에 코드를 쫓아가기가 쉽지않다?

```swift
func makeJuice(_ juice: Juice) {
    if checkFruitInJuice(juice) {
        useFruit(juice)
    }
}
    
private func checkFruitInJuice(_ juice: Juice) -> Bool {
    for (fruit, num) in juice.recipe {
        do {
            try fruitStore.check(fruit: fruit, by: num)
        } catch FruitStoreError.noExistInventory {
            print("해당 과일 존재하지 않음")
            return false
        } catch FruitStoreError.lackedInventory {
            print("과일 재고 부족")
            return false
        } catch {
            print(error)
            return false
        }
    }
    return true
}
```
* 위는 에러처리를 ```checkFruitInJuice```에서 사용하는 방법입니다.
* 장점 - ```FruitStore```의 에러가 한번만 이동하여 처리 된다.
* 단점 - ```makeJuice```가 너무 단순화해지며, 에러가 났을 경우 다양한 에러처리 방법이 어렵다.
* 양쪽 다 동작도 되며, 에러처리문도 사용이 가능하지만 각자 장단점이 있고 어떤것이 더 사용하기 괜찮은 방법일지 잘 모르겠습니다😅😅

### 매직 넘버
* 각 과일의 수량은 10이라는 숫자로 초기화 해줍니다. 
```swift
private var fruits: [Fruit: Int] = [.strawberry: 10, .banana: 10, .kiwi: 10, .mango: 10, .pineapple: 10]
```
* 코드를 짤 때 각각 할당해주는 것이 지금은 문제가 없으나 나중에 만약이라도 초기 수량을 바꾸게 된다면 하나하나 수정을 해야한다는 문제가 발생합니다.
* 이러한 문제점을 해결하는 좋은 방법이 있는지 모르겠습니다.

### 캡슐화, 은닉화

#### 은닉화
* 은닉화에 대해서 많이 고민했습니다. 은닉화란 내부 데이터 또는 내부 메서드를 접근하지 못하게 만드는 것인데 은닉화를 생각하니 `FruitStore` 클래스 안에 들어있는 `fruits`라는 프로퍼티가 외부로 들어나면 안된다고 생각하여 `private`으로 접근 제어를 하였습니다.
* 마찬가지로 `JuiceMaker` 구조체 안에 들어있는 `fruitStore`라는 instance도 `private`으로 접근 제어를 하였습니다.
```swift
class FruitStore {
    private var fruits: [Fruit: Int] = [.strawberry: 10, .banana: 10, .kiwi: 10, .mango: 10, .pineapple: 10]
}

struct JuiceMaker {
    private let fruitStore = FruitStore()
}
```
* 은닉화를 외부로 들어나서는 안되는것, 외부에서 사용하지 않는 것으로 기준을 삼아서 숨겼습니다. 은닉화의 기준을 이렇게 잡고 코드를 작성해도 되는걸까요?


#### 캡슐화
* 캡슐화란 ```객체의 속성(data fields)과 행위(메서드, methods)를 하나로 묶고, 실제 구현 내용 일부를 내부에 감추어 은닉한다.``` 이렇게 말하고는 하는데 그러면 저희가 만든 JuiceMaker는 안에 fruitStore과 makeJuice등 프로퍼티와 메소드를 구조체로 하나를 묶고 은닉화한것을 캡슐화라고 말을 할 수 있을까요?
```swift
func makeJuice(_ juice: Juice) {
    if checkFruitInJuice(juice) {
        useFruit(juice)
    }
}
```
* 추가로 위 함수처럼 특정 메소드를 불러와서 하나의 함수로 만든것 역시 캡슐화라고 말할 수 있을까요?

fatherLeon committed:
chore: 주석, 함수 생성

주석내용변경 및 increaseFruit, decresaeFruit 함수 선언

chore

Commit:
feat: 열거형 생성

Fruit열거형 생성

feat

fatherLeon committed:
feat: 함수 기능 완료

increaseFruit, decresaeFruit 함수 구현

feat

Commit:
feat: 열거형 함수 구현

error타입 열거형 생성, 재고확인 함수 구현

feat

fatherLeon committed:
feat: 열거형 생성, 열거형 기능 추가

JuiceMenu 생성, Fruit 변수 추가

feat

Commit:
feat: makeJuice함수 생성

fatherLeon committed:
feat: JuiceMenu 열거형 레시피 구현

Commit:
feat: makeJuice함수 구현

fatherLeon committed:
feat: 열거형 케이스 생성, 함수 수정

JuikceError noExistInventory케이스 생성, checkFruit 에러처리 함수로 변경

feat

Commit:
feat: makeJuice함수 로직 변경, error열거형 이름 변경

fatherLeon committed:
feat: 타입에 의한 파일 분리

fatherLeon committed:
chore: viewcontroller 테스트

fatherLeon committed:
feat: 함수 메소드 구현

checkFruitInJuice, useFruit 메소드 구현, makeJuice 수정

feat

Commit:
feat: 함수 로직 변경

makeJuice함수 로직 변경, checkFruitInJuice 로직 변경

feat

fatherLeon committed:
style: 함수 이름과 인자 변경, 은닉화

FruitStore -> fruit은닉화 및 increase, decrease, check 함수명 변경 및 함수 인자 수정
JuiceMaker -> fruitStore, checkFruitJuice, useFruit 은닉화 및 JuiceMaker내 함수 메소드 라벨 수정

style

fatherLeon committed:
style: checkFruitInJuice -> checkFruitInStore 함수명 변경

seu0313 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,21 @@
+//  Created by Vetto, 레옹아범 on 2023.1.2
+
+enum Fruit {
+    case strawberry, banana, kiwi, mango, pineapple
```
열거형의 case를 한 줄로 표현해주셨네요 👍

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
-
-// 과일 저장소 타입
 class FruitStore {
+    private var fruits: [Fruit: Int] = [.strawberry: 10, .banana: 10, .kiwi: 10, .mango: 10, .pineapple: 10]
+    
```
[P2] 이 코멘트는 질문주신 부분과 연결되는 것 같네요. FruitStore가 init() 메서드를 통해 초기화 되는 시점에서 설정해주면 해결할 수 있을 것 같아요.

fatherLeon replied:
```swift
class FruitStore {
    init(strawberry: Int, banana: Int, kiwi: Int, mango: Int, pineapple: Int) {
        self.fruits[.strawberry] = strawberry
        self.fruits[.banana] = banana
        self.fruits[.kiwi] = kiwi
        self.fruits[.mango] = mango
        self.fruits[.pineapple] = pineapple
    }
    
    convenience init(amount: Int) {
        self.init(strawberry: amount, banana: amount, kiwi: amount, mango: amount, pineapple: amount)
    }
}
```
* 이렇게 이니셜라이저 2개를 작성하였습니다.
* 우선 첫번째 이니셜라이저의 경우 과일마다 따로 값을 받고 싶을 경우 사용을 하기위해 만들었습니다.
* 두번째의 경우 convenience 이니셜라이저 구문은 편의성을 위한 구문으로 이해했으며, 같은 객체내의 class에 구현된 이니셜라이저를 가져와서 작성해야한다고 공식문서에 써져있어서 하나의 값으로 전부 다 값을 초기화 시켜주기 위해서 사용했습니다.
* convenience구문을 사용한 가장 큰 이유는 단순하게 init을 사용할 경우 딸기부터 파인애플까지 총 5개의 값을 초기화 시켜줘야하지만 convenience구문을 사용하면 이미 만들어져있는 init을 활용하여 코드도 짧고 보기 쉽게 이해할 수 있기 때문에 사용했습니다.

seu0313 replied:
좋습니다 👍

seu0313 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,31 @@
+    case strawberryBananaJuice
+    case mangoKiwiJuice
+    
+    var recipe: [Fruit: Int] {
+        switch self {
+        case .strawberryJuice:
+            return [.strawberry: 16]
+        case .bananaJuice:
+            return [.banana: 2]
+        case .kiwiJuice:
+            return [.kiwi: 3]
+        case .pineappleJuice:
+            return [.pineapple: 2]
+        case .mangoJuice:
+            return [.mango: 3]
+        case .strawberryBananaJuice:
+            return [.strawberry: 10, .banana: 1]
+        case .mangoKiwiJuice:
+            return [.mango: 2, .kiwi: 1]
```
딕셔너리를 사용해서 레시피를 저장하신 부분 아주 좋네요 :)

seu0313 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,31 @@
+//  JuiceMaker - JuiceMenu.swift
```
[P3] 주석 내용과 실제 타입 이름이 다른 것 같아요.

gzzjk159 replied:
나중에 JuiceMenu보다 그냥 Juice라고 하는 것이 더 자연스러워 보여서 바꾸었는데 헤더 주석은 깜빡하고 바꾸지 못했던거 같습니다. 해당 내용은 파일 이름에 맞게 수정하였습니다.  2848614e4799491370867334df71fccb54adcd01

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
 class FruitStore {
+    private var fruits: [Fruit: Int] = [.strawberry: 10, .banana: 10, .kiwi: 10, .mango: 10, .pineapple: 10]
+    
+    func increase(fruit: Fruit, by amount: Int) {
```
[P3] `func increaseFruit(_ fruit: Fruit, by amount: Int)`는 어떠신가요?
현재는 `makeJuice(_ juice: Juice)` 에서 하셨던 네이밍과 다른 것 같아요

gzzjk159 replied:
통일성이 없어보이는 naming 이었던거 같습니다. 해당 문제점을 파악하고 상의하여 리뷰어가 조언해주신 increaseFruit로 함수명을 변경하여 가독성을 높였습니다. 또한 fruit 매개변수 앞에 _를 붙여서  Argument Label 을 생략하여 호출할때 전달인자 생략을 할 수 있게 하였습니다. 2848614e4799491370867334df71fccb54adcd01

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,36 @@
-//  Created by yagom. 
-//  Copyright © yagom academy. All rights reserved.
-// 
+//  Created by Vetto, 레옹아범 on 2023.1.2
```
주석을 변경해주셨네요 좋습니다 :)

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,36 @@
+        for (fruit, num) in juice.recipe {
+            do {
+                try fruitStore.check(fruit: fruit, by: num)
+            } catch FruitStoreError.noExistInventory {
+                print("해당 과일 존재하지 않음")
+                return false
+            } catch FruitStoreError.lackedInventory {
+                print("과일 재고 부족")
+                return false
+            } catch {
+                print(error)
+                return false
```
[P2] 각각의 에러를 추출해서 출력문을 작성하는 것보다 LocalizedError 프로토콜을 FruitStoreError에 채택하여 사용하는 것이 좋을 것 같아요

```swift
enum FruitStoreError: LocalizedError { ... }

do {
  ...

} catch {
  print(error.localizedDescription)
}
```

fatherLeon replied:
```swift
enum FruitStoreError: Error, LocalizedError {
    case lackedInventory(fruit: Fruit, lackedAmount: Int)
    case noExistInventory
    
    var errorDescription: String? {
        switch self {
        case .lackedInventory(let fruit, let amount):
            return "\(fruit) 과일 \(amount)개 부족"
        case .noExistInventory:
            return "과일 존재하지 않음"
        }
    }
}

private func checkFruitInStore(_ juice: Juice) -> Bool {
    for (fruit, num) in juice.recipe {
        do {
            try fruitStore.checkFruit(fruit, by: num)
        } catch let error as FruitStoreError {
            if let message = error.errorDescription {
                print(message)
            }
            return false
        } catch {
            print(error)
            return false
        }
    }
    return true
}
```
* 우선은 이렇게 작성했습니다.
* 다만, 저희가 질문을 잘 이해한건지 모르겠고 Error에 enum에 추가적인 질문이 있어서 밑에 PR코멘트에 추가로 작성하도록 하겠습니다

seu0313 replied:
LocalizedError는 내부적으로 Error 프로토콜을 채택하고 있습니다 LocalizedError만 채택해줘도 될 것 같아요 👍 

![스크린샷 2023-01-04 오후 6 55 11](https://user-images.githubusercontent.com/94151993/210529438-2afdef04-bde0-41cb-8745-649d291d6e0d.png)

seu0313 replied:
LocalizedError를 채택했고 errorDescription 을 정의해주었기 때문에 아래와 같이 줄일 수 있을 것 같아요.

```swift
private func checkFruitInStore(_ juice: Juice) -> Bool {
    for (fruit, num) in juice.recipe {
        do {
            try fruitStore.checkFruit(fruit, by: num)
        } catch {
            print(error.localizedDescription)
            return false
        }
    }
    return true
}
```

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,36 @@
+        return true
+    }
     
+    private func useFruit(_ juice: Juice) {
```
[P1] Argument Label을 사용하여 네이밍을 변경해보면 좋을 것 같습니다 :)

fatherLeon replied:
```swift
private func useFruit(in juice: Juice) {
}
```
* 위와 같이 in을 사용하여 쥬스안의 과일을 사용한다라는 식의 네이밍으로 변경하였습니다!!

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,7 @@
+//  JuiceMaker - FruitStoreError.swift
+//  Created by Vetto, 레옹아범 on 2023.1.2
+
+enum FruitStoreError: Error {
+    case lackedInventory
+    case noExistInventory
+}
```
[P3] LocalizedError에 대해 한번 공부해보시면 좋을 것 같아요. 에러가 발생했을때 출력할 Description를 에러 타입에 정의해둔 후 사용한다면 새로운 에러 케이스가 추가될 때 출력문 코드를 찾아가지 않아도 되는 장점이 있습니다.

gzzjk159 replied:
LocalizedError: `A specialized error that provides localized messages describing the error and why it occurred.`
오류가 발생한 이유에 대해서 설명하는 지역화된 메세지를 제공하는 특수한 오류라고 나옵니다.

사용 예시들을 찾아보며 해당 에러처리 부분을 수정하였습니다. 
그 결과 에러가 발생했을때 에러 하나하나 print하는 것이 아닌 에러를 찾고 에러처리 안에서 에러메세지를 받고 출력하는 방법으로 수정하였습니다. 
1차 수정 : a742630c2f7b9dca8289f2a338132cf8a158f9fa
2차 수정 : 5e1598663dc61139b0f64bb41a6fc1b9c764a3b2 -> 타입 캐스팅 사용, 에러 string 상세화

fatherLeon replied:
> https://developer.apple.com/documentation/foundation/localizederror
* 에러가 나타났을 경우 에러에 대한 설명을 구체적으로 할 수 있게 도와주는 프로토콜인 것 같습니다.

#### 네가지의 프로토콜 인스턴스 프로퍼티
* errorDescription - 에러에 관한 설명
* failureReason - 에러가 뜬 실패의 이유
* helpAnchor - 번역하면 ```닻을 내리다```로 해석이 되는데 이부분은 에러 발생시 해결할 수 있는 방안에 대해 설명해주는 프로퍼티라고 이해했습니다. 
* recoverySuggestion - 이 부분도 helpAnchor와 마찬가지로 에러를 해결 할 수 있는 부분에 대해 제안을 해주는 프로퍼티로 이해했는데 helpAnchor와 의미가 유사해서 조금 더 공부하도록 하겠습니다.

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
+        self.fruits[fruit] = storedFruit + amount
+    }
+    
+    func decrease(fruit: Fruit, by amount: Int) {
+        guard let storedFruit = self.fruits[fruit] else {
+            return
+        }
+        self.fruits[fruit] = storedFruit - amount
+    }
     
+    func check(fruit: Fruit, by amount: Int) throws {
+        guard let storedFruit = self.fruits[fruit] else {
+            throw FruitStoreError.noExistInventory
+        }
+        guard storedFruit >= amount else {
+            throw FruitStoreError.lackedInventory
+        }
```
[P3] 위의 increase 코멘트의 내용과 동일합니다.

gzzjk159 replied:
위에서 말한대로 다른 코드랑 같이 변경을 하여 통일성을 주고 가독성을 높였습니다. 2848614e4799491370867334df71fccb54adcd01

seu0313 reviewed (changes requested):
안녕하세요 vetto, 레옹아범
쥬스메이커 프로젝트 STEP 1 수고하셨습니다 :)

---

앞으로의 원활한 리뷰 진행을 위해 몇 가지 안내사항을 드리겠습니다.

1. **PR을 보내기 전에는 3가지를 확인해주세요.**
    - `캠핑사이트 - Course - Pull Request 보낼 때` 글을 꼭 확인해주세요.
    - `yagom-academy:main`로 보내고 있는 것은 아닌 지 확인해주세요.
    - 이전 STEP이 Merge된 후 다음 STEP의 PR을 보내주세요.
2. **PR 수정 코멘트에는 P1 - P3의 중요도를 남겨두었습니다. 중요하지 않는 경우 확인했다는 이모지만 남기셔도 됩니다.**
    - [P1] 무조건 수정이 필요합니다.
    - [P2] 다음 스탭에서 수정해도 됩니다.
    - [P3] 꼭 반영하지 않고 참고만 해도 됩니다.
3. **PR 처럼 수정 후에도 DM를 남겨주시면 빠르게 확인하겠습니다.**

---

**Q1. 소유에 대한 의미**
> 이번 프로젝트 요구사항 중에는 JuiceMaker는 FruitStore를 소유하고 있습니다.라는 사항이 있었습니다. 처음에는 어떠한 의미인지 모르겠어서 고민을 많이 했습니다. ... 저희가 생각하고 있는 것이 맞는지 궁금합니다.

A1. 네 현재 구현해주신대로 JuiceMaker는 FruitStore를 소유하고 있기 때문에 
JuiceMaker는 FruitStore를 의존하고 있다라고도 할 수 있겠네요.


**Q2. 에러처리에 대한 로직**
> throw를 두번 거치는 방법에 대해 어떻게 생각하시나요?

A2. 개인적인 의견으로는 적절한 선에서는 나쁘지 않다고 생각합니다. 결과적으로 해당 메서드 호출의 결과로 에러가 발생했을 때 처리해야하는 곳에서 처리만 된다면 괜찮다고 생각합니다.

**Q3. 매직 넘버**
> 코드를 짤 때 각각 할당해주는 것이 지금은 문제가 없으나 나중에 만약이라도 초기 수량을 바꾸게 된다면 하나하나 수정을 해야한다는 문제가 발생합니다. 이러한 문제점을 해결하는 좋은 방법이 있는지 모르겠습니다.

A3. 해당 질문은 코멘트에 남겨드렸습니다 확인 부탁드립니다 :)

**Q4. 캡슐화, 은닉화**
> 은닉화를 외부로 들어나서는 안되는것, 외부에서 사용하지 않는 것으로 기준을 삼아서 숨겼습니다. 은닉화의 기준을 이렇게 잡고 코드를 작성해도 되는걸까요? 추가로 위 함수처럼 특정 메소드를 불러와서 하나의 함수로 만든것 역시 캡슐화라고 말할 수 있을까요?

A4. 은닉화, 캡슐화 모두 잘 적용해주셨습니다. 외부에서는 `makeJuice(_:)`라는 메서드를 호출했을때 내부에서 어떠한 동작이 있는 지를 모르고 결과적으로 Juice를 만들어 줄 것이라고 기대할 수 있을 것 같아요. 개인적인 의견으로는 캡슐화라고 볼 수 있다라고 생각합니다.

Commit:
feat: init 작성

FruitStore initialize 작성

feat

fatherLeon committed:
style: 코드함수명 수정

Commit:
feat: 에러처리 출력 로직 변경

FruitStoreError LocalizedError 프로토콜 채택, JuiceMaker에러 출력 변경

feat

fatherLeon committed:
feat: 에러관련 내용 수정

JuiceMaker의 에러처리 catch문 두개로 변경
FruitStoreError lackedInventory 과일명과 모자른 과일 개수까지 출력

feat

fatherLeon committed:
chore: 테스트용 print문 삭제

fatherLeon committed:
chore: 변수 private추가

fatherLeon committed:
chore: FruitStore의 fruits

fatherLeon commented:
@llingo 
안녕하세요 링고 보내주신 피드백 많은 도움이 됐습니다
피드백을 통해 공부하고 고치다보니 몇가지 질문점이 있어서 추가로 질문드립니다~~
좋은 피드백 감사합니다👍

## 추가로 궁금한 점 && 조언 받고 싶은 점

### Extension 활용하는 Error처리
* 아래 차이점말고도 차이가 더 있을까요?
    * code 1에서는 error문과 localizederror문이 같이 있어서 어떤 오류가 있는지 그 오류에 관하여 나오는 메세지는 무엇인지 한번에 확인할 수 있는 장점이 있지만 수정하게 되면 같이 수정해야한다는 단점이 있습니다. 
    * code 2에서는 한번에 파악하긴 어렵다는 단점이 있지만 리팩토링을 할 때 그 부분만 수정하면 되는 장점이 있습니다.
* 밑의 예제처럼 프로토콜을 여러개 채택할 경우 각각에 대한 프로토콜의 필수 혹은 선택적 부분을 작성해 가독성을 높여주기 위해 extension을 code2처럼 사용한다고 생각합니다, 그런데 저희 FruitStoreError와 같이 적은 코드 수를 가진 경우도 프로토콜을 다중채택하면 extension을 사용해주는게 맞는 판단일까요?

> code 1
```swift
enum FruitStoreError: Error, LocalizedError {
    case lackedInventory(fruit: Fruit, lackedAmount: Int)
    case noExistInventory
    
    var errorDescription: String? {
        switch self {
        case .lackedInventory(let fruit, let amount):
            return "\(fruit) 과일 \(amount)개 부족"
        case .noExistInventory:
            return "과일 존재하지 않음"
        }
    }
}
```

> code 2
```swift
enum FruitStoreError: Error {
    case lackedInventory(fruit: Fruit, lackedAmount: Int)
    case noExistInventory
}

extension FruitStoreError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .lackedInventory(let fruit, let amount):
            return "\(fruit) 과일 \(amount)개 부족"
        case .noExistInventory:
            return "과일 존재하지 않음"
        }
    }
}
```



### 에러처리구문
```swift
private func checkFruitInStore(_ juice: Juice) -> Bool {
    for (fruit, num) in juice.recipe {
        do {
            try fruitStore.checkFruit(fruit, by: num)
        } catch let error as FruitStoreError {
            if let message = error.errorDescription {
                print(message)
            }
            return false
        } catch {
            print(error)
            return false
        }
    }
    return true
}
```
* 이 부분에서 do - catch문 말씀하신게 catch문을 하나만 작성하라는 말씀이셨을까요?
* 아니면 기존에 3개의 catch문 중 2개는 FruitStoreError라는 에러타입이 같기 때문에 catch문을 하나로 통합하라는 말씀이셨을까요?
```swift
do {
    try fruitStore.checkFruit(fruit, by: num)
} catch {
    switch error {
    case let error as FruitStoreError:
        print(error.errorDescription)
    case error:
        print(error.localizedDescription)
    }
    return false
}
```

* 위 코드는 catch문 하나만 작성한 코드입니다.
* catch문 하나만 작성하라는 의미셨으면 코드도 짧아져 가독성이 좋아서 제안을 주신거라고 생각합니다.
* 저희 생각으로는 FruitStoreError의 종류가 더 늘어나고 각각의 에러 케이스에 따라 처리도 필요할것 같아서 우선은 위 FruitStoreError에 관한 catch문 하나, 일반적인 error의 catch문 하나로 작성하였습니다.!!!

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,41 @@
-
-// 과일 저장소 타입
 class FruitStore {
+    private var fruits: [Fruit: Int] = [.strawberry: 0, .banana: 0, .kiwi: 0, .mango: 0, .pineapple: 0]
+    
+    init(strawberry: Int, banana: Int, kiwi: Int, mango: Int, pineapple: Int) {
+        self.fruits[.strawberry] = strawberry
+        self.fruits[.banana] = banana
+        self.fruits[.kiwi] = kiwi
+        self.fruits[.mango] = mango
+        self.fruits[.pineapple] = pineapple
+    }
+    
+    convenience init(amount: Int) {
+        self.init(strawberry: amount, banana: amount, kiwi: amount, mango: amount, pineapple: amount)
+    }
```
convenience init을 잘 활용해주셨습니다. 👍 
하지만 만약 과일 종류가 더 늘어난다면 생성자의 매개변수가 하나 더 추가되는 문제가 있을 것 같아요.

[P3] 만약, 현재처럼 생성자를 통해 생성해주려면 다음과 같은 방식으로 생성해줘도 좋을 것 같아요 😄

```swift
struct 과일재고정보 {
  let strawberry: Int
  let banana: Int
  let kiwi: Int
  let mango: Int
  let pineapple: Int
}

init(재고정보: 과일재고정보) {
    self.fruits[.strawberry] = 재고정보.strawberry
    self.fruits[.banana] = 재고정보.banana
    self.fruits[.kiwi] = 재고정보.kiwi
    self.fruits[.mango] = 재고정보.mango
    self.fruits[.pineapple] = 재고정보.pineapple
}
```

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+enum FruitStoreError: Error, LocalizedError {
```
[P2] LocalizedError는 Error를 채택한 프로토콜입니다. 현재는 중복 채택되었다고 볼 수 있을 것 같아요.

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
+        for (fruit, num) in juice.recipe {
+            do {
+                try fruitStore.checkFruit(fruit, by: num)
+            } catch let error as FruitStoreError {
+                if let message = error.errorDescription {
+                    print(message)
+                }
+                return false
+            } catch {
+                print(error)
+                return false
+            }
```
[P3] error.localizedDescription을 사용하면 더 간단하게 줄일 수 있을 것 같아요 :)

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
+        return true
+    }
     
+    private func useFruit(in juice: Juice) {
```
Argument Label을 사용해주셨네요👏

seu0313 reviewed (approved):
수고하셨습니다 :) 잘 수정해주셨네요
몇 가지 추가 피드백의 경우 다음 스탭에서 수정하셔도 될 것 같아요. 
다음 스탭 PR 전에 코멘트 확인 부탁드립니다. :)

**Q1. Extension 활용하는 Error처리**
> 밑의 예제처럼 프로토콜을 여러개 채택할 경우 각각에 대한 프로토콜의 필수 혹은 선택적 부분을 작성해 가독성을 높여주기 위해 extension을 code2처럼 사용한다고 생각합니다, 그런데 저희 FruitStoreError와 같이 적은 코드 수를 가진 경우도 프로토콜을 다중채택하면 extension을 사용해주는게 맞는 판단일까요?

A1. LocalizedError는 내부에 Error 프로토콜을 채택하고 있습니다!

저의 경우에는 간단한 경우에는 하나의 블럭 안에 넣지만 만약, 중첩 타입을 사용하는 등 블럭 내부에 코드가 길어지는 경우에는 가독성을 위해 extension으로 errorDescription 부분을 분리해서 사용하고 있습니다.

```swift
// 예시
enum RingoError: LocalizedError {
    enum firstReason {
      case reason1
      case reason2
     }

    enum secondReason {
      case reason1
      case reason2
     }
    
    case first(reason: firstReason)
    case second(reason: secondReason)
}


// MARK: Extension

extension RingoError {
    var errorDescription: String? {
      ...
    }
}
```

**Q2. 에러처리구문**
A2. 코멘트에서도 남겨 드렸지만 LocalizedError를 채택하였고 errorDesciption를 정의해두었기 때문에 아래와 같이 간단하게 처리할 수 있을 것 같아요 :)

```swift
private func checkFruitInStore(_ juice: Juice) -> Bool {
    for (fruit, num) in juice.recipe {
        do {
            try fruitStore.checkFruit(fruit, by: num)
        } catch {
            print(error.localizedDescription)
            return false
        }
    }
    return true
}
```

다음 스탭도 화이팅입니다. 🎉
